Recommend App: Add presenter to encapsulate display logic

dvdchr opened the pull request:
Refs #16995
Closes #17003
Based on #17005 

## Summary
This change introduces a presenter object that encapsulates content fetching and presentation logic for the share app flow, along with a subclass of `UIActivityItemSource` to help format the text contents depending on the activity type. As described in #17003, I think it's worth it to encapsulate this since it's possible to have multiple entry points in the future.

Some notes for this change:
- The presenter exposes a delegate that gets called on every loading state changes. This is useful for the caller to update their display to show e.g. a loading indicator while the content is being fetched.
- Other than formatting needs, a `UIActivityItemSource` subclass is necessary to provide additional information – specifically a subject string provided through `activityViewController:subjectForActivityType`. This information is used when sharing to mail. I'll share some screenshots below on how it looks when it's shared in various apps.
- Lastly, this adds a new feature flag for the recommend button.

## Screenshots

Share Destination | Screenshot
--|--
Messages | ![share_app_messages](https://user-images.githubusercontent.com/1299411/129242722-dca2fbd5-6ca5-43df-b89b-e9937708d1ec.jpeg)
Mail | ![share_app_mail](https://user-images.githubusercontent.com/1299411/129241907-c3234192-8459-4eb5-b1ab-e07951b3cb4c.jpeg)
Twitter | ![share_app_twitter](https://user-images.githubusercontent.com/1299411/129241719-8e19e3f8-e95f-4f7f-9187-873032ec32cc.jpeg)
WordPress (Save as Draft) – **Note:** Not sure why the text contents are not carried, but I suspect this needs to be fixed from the WP side. | ![share_app_wp_draft](https://user-images.githubusercontent.com/1299411/129241721-af57f4d9-0cff-4937-bc10-f0b2460ccfbc.png)
Copy (Pasteboard) | ![share_app_copyToPasteboard](https://user-images.githubusercontent.com/1299411/129241726-61316db5-95a5-4351-83e3-6a64d9a308a5.jpg)
Chrome | ![share_app_chrome](https://user-images.githubusercontent.com/1299411/129241730-c37b21be-b74d-4c0e-81ef-b9f7b1c6d906.png)
Slack – Not sure why the text is not present, but this is usually due to the destination's Share Extension not picking up multiple activity items. | ![share_app_slack](https://user-images.githubusercontent.com/1299411/129241736-953a0f51-780f-4f2d-aafc-cb6909dd9d0e.png)
WhatsApp | ![share_app_whatsapp](https://user-images.githubusercontent.com/1299411/129241740-dde4d0c1-230a-4d34-8b2b-7313c968cc6a.png)

## To Test

### Manually editing the source + compiling to device
Since there are no visual components hooked up to this component, we'll have to wire this up manually in Xcode AND compile it to a device 😅 . 

Here's how I modified the code to test the share functionality:

- In the `MeViewController`, add a private instance variable like so:
  ```Swift
  private let sharePresenter = ShareAppContentPresenter(account: nil)
  ```
- Find the `pushHelp()` method, comment out the code in the return block, and change it to this:
  ```Swift
  return { [unowned self] row in
      defer { self.tableView.deselectSelectedRowWithAnimation(true) }
      guard let indexPath = self.tableView.indexPathForSelectedRow,
            let selectedCell = self.tableView.cellForRow(at: selectedIndexPath) else {
          return
      }
      sharePresenter.present(for: .wordpress, in: self, sourceView: selectedCell)
  }
  ```
- Since there's no loading indicator yet, just wait for one or two seconds and the share sheet should appear shortly after tapping the help button on the Me screen.
- Lastly, try logging in to a self-hosted site (without WP account) and make sure that you can still prompt the share sheet.

### Unit tests
Make sure that the tests are 🟢  !

## Regression Notes
1. Potential unintended areas of impact
None.

2. What I did to test those areas of impact (or what existing automated tests I relied on)
This component is not used yet.

3. What automated tests I added (or what prevented me from doing so)
Added some unit tests for the presenter.

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

dvdchr committed:
Add ShareAppContentPresenter and custom UIActivityItemSource

dvdchr committed:
Add feature flag for recommend app

dvdchr committed:
Add tests for presenter

dvdchr committed:
Improve test speed

dvdchr committed:
Show the text first before the URL

dvdchr committed:
Allow the presenter to function with no WPAccount

dvdchr committed:
Merge branch 'issue/17002-button-with-icon-cell' into issue/17003-share-app-content-presenter

dvdchr committed:
Merge branch 'issue/17002-button-with-icon-cell' into issue/17003-share-app-content-presenter

dvdchr committed:
Add tests for the activity item source

dvdchr committed:
Allow initializing with nil, do not perform request when loading

dvdchr committed:
Set feature flag value to false

dvdchr commented on WordPress/Classes/ViewRelated/Sharing/ShareAppTextActivityItemSource.swift:
```diff
@@ -0,0 +1,36 @@
+// MARK: - Localized Strings
+
+private extension String {
+    static let defaultSubjectText = NSLocalizedString("WordPress.com Apps - Apps for any screen",
```
@aerych, @mattmiklic are we good to go with this subject text? This text is used as the email subject line when sharing the app via Mail.app.

(Also, worth noting that Gmail.app somehow ignores this value, and takes the text content instead. I'm suspecting it's a bug from their side. 😅 )

mattmiklic replied:
This is fine with me unless @aerych wants to wordsmith it. 😄

aerych replied:
Maybe `WordPress Apps` rather than `WordPress.com Apps`?  If we omit the .com it might sound a tad more inclusive for self-hosted/Jetpack sites.

dvdchr replied:
Addressed in https://github.com/wordpress-mobile/WordPress-iOS/pull/17006/commits/b963748036929e771fd66ee58a740f63515e84ef.

dvdchr committed:
Add files to Jetpack. It was not added somehow

aerych reviewed (approved):
Howdy @dvdchr 👋 
I tested on my iPhone X and everything worked great.  Thanks for showing all the examples in the PR description. It was nice to have something to compare my results to. :) 
Code looks good to me. I had just one suggestion about the subject text. 
Otherwise :shipit: :)

aerych commented on WordPress/Classes/ViewRelated/Sharing/ShareAppContentPresenter.swift:
```diff
@@ -0,0 +1,121 @@
+        ]
+
+        let activityViewController = UIActivityViewController(activityItems: activityItems, applicationActivities: nil)
+        viewController.present(activityViewController, animated: true, completion: nil)
```
I approved too soon 😭.  I neglected to test on the iPad and went back to do that.  On the iPad the activity controller is presented as a popover.  We'll need to define a source view or source rect for the activityViewController's popoverPresentationViewController or else we'll encounter an exception: 
> Terminating app due to uncaught exception 'NSGenericException', reason: 'UIPopoverPresentationController (<UIPopoverPresentationController: 0x7ff5e70296a0>) should have a non-nil sourceView or barButtonItem set before the presentation occurs.'

Something like `activityViewController.popoverPresentationController?.sourceView = viewController.view ` might be good enough. (Probably needs to be before the call to `.present`.)

dvdchr replied:
Thank you for noticing this. I completely missed this one! This is addressed in https://github.com/wordpress-mobile/WordPress-iOS/pull/17006/commits/a9bb473e77a0c7a120ff11fe5475eae7e5d357f6. Tested in iPad simulator and it now works as expected.

aerych reviewed (changes requested):
Spotted an issue so I've added another note and marked this request changes since we need to address an exception.  Sorry I missed it in the first round!

dvdchr committed:
Assign sourceView for iPad presentation context

dvdchr committed:
Updated copy for share app subject text

dvdchr committed:
Adjust tests for share app activity item source

dvdchr committed:
Merge remote-tracking branch 'origin/develop' into issue/17003-share-app-content-presenter

aerych reviewed (approved):
Heya @dvdchr 👋 
The exception is resolved. Thanks for the changes! Marking approved so as to not block the rest of the work.  

However, I think we might want to see if we can improve the popover presentation on iPad. The solution I suggested was something I hadn't tested, and now that I see it in action I think we might need to make an adjustment.   

Depending on which way I orient the iPad, the popover appears away from where I tapped. The size also varies depending on orientation so sometimes there are more visible options than others.  I'm wondering if sharing the frame of the cell (or maybe just the cell as the view) that was tapped rather than the controller view would help improve the experience.  We'd also need to make the change before the code freeze.  What do you think? 
![Simulator Screen Shot - iPad Pro (9 7-inch) - 2021-08-18 at 11 50 02](https://user-images.githubusercontent.com/1435271/129940260-a4f1f76a-08cc-44e2-b1eb-9f51df6d023f.png)
![Simulator Screen Shot - iPad Pro (9 7-inch) - 2021-08-18 at 11 50 05](https://user-images.githubusercontent.com/1435271/129940262-82ce5ba1-41d9-41c0-b9ca-dc2515772307.png)

dvdchr committed:
Add sourceView parameter for proper popover anchoring

dvdchr commented:
Hey @aerych. I wasn't very aware of the design guidelines for iPad before, so thank you for the help and tips on this. Really appreciate it! 🙂

> Depending on which way I orient the iPad, the popover appears away from where I tapped. The size also varies depending on orientation so sometimes there are more visible options than others. I'm wondering if sharing the frame of the cell (or maybe just the cell as the view) that was tapped rather than the controller view would help improve the experience.

Noted. I just realized that since the popover is anchored to the view controller's view, there's not enough space on the side and the system puts the popover on the top side, which makes the content a bit cramped. 

Addressed in https://github.com/wordpress-mobile/WordPress-iOS/pull/17006/commits/6a07b6d63227fdc970c3be61220d54903d9e5fae, I've confirmed that passing the cell as the `sourceView` correctly anchors the popover to the tapped cell, and indeed the share sheet is displayed better. Here's a preview:

Portrait | Landscape
--|--
![ipad_portrait](https://user-images.githubusercontent.com/1299411/130045079-e2a7094b-4ad4-4eb0-8f13-1eacaf0125c6.png) | ![ipad_landscape](https://user-images.githubusercontent.com/1299411/130045091-7f5fa225-7a1d-4a64-9e86-02c89f0e63b2.png)

> We'd also need to make the change before the code freeze. What do you think?

Since the code review for #17007 and #17020 is blocked by this one, I'll apply the changes on this PR. Then, I'll merge up the changes to both #17007 and #17020, so you can proceed with reviewing both of them. 

I'll aim to merge everything to `develop` tomorrow (provided that you approved them all, of course 😄 ). If there's any new major issue on the PRs, I will address it on a separate PR (since the feature flag is still `false`).

---
P.S.: I've updated the code sample in the description to fill the `sourceView` parameter if you decided to give this another look! 🙂

aerych commented:
Tested with the changes and they look great!  Thanks @dvdchr!  :shipit: again!

dvdchr commented:
Thanks @aerych !
